pkg/cli/admin/mustgather/mustgather.go

```diff
@@ -412,64 +479,17 @@ func (o *MustGatherOptions) Run() error {
 	defer o.logTimestamp()
 
 	var wg sync.WaitGroup
-	wg.Add(len(pods))
 	errCh := make(chan error, len(pods))
-	for _, pod := range pods {
-		go func(pod *corev1.Pod) {
-			defer wg.Done()
-
-			if len(o.RunNamespace) > 0 && !o.Keep {
-				defer func() {
-					// must-gather runs in its own separate namespace as default , so after it is completed
-					// it deletes this namespace and all the pods are removed by garbage collector.
-					// However, if user specifies namespace via `run-namespace`, these pods need to
-					// be deleted manually.
-					err = o.Client.CoreV1().Pods(o.RunNamespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})
-					if err != nil {
-						klog.V(4).Infof("pod deletion error %v", err)
-					}
-				}()
-			}
-
-			log := newPodOutLogger(o.Out, pod.Name)
+	podChan := make(chan *corev1.Pod)
 
-			// wait for gather container to be running (gather is running)
-			if err := o.waitForGatherContainerRunning(pod); err != nil {
-				log("gather did not start: %s", err)
-				errCh <- fmt.Errorf("gather did not start for pod %s: %s", pod.Name, err)
-				return
-			}
-			// stream gather container logs
-			if err := o.getGatherContainerLogs(pod); err != nil {
-				log("gather logs unavailable: %v", err)
-			}
-
-			// wait for pod to be running (gather has completed)
-			log("waiting for gather to complete")
-			if err := o.waitForGatherToComplete(pod); err != nil {
-				log("gather never finished: %v", err)
-				if exiterr, ok := err.(*exec.CodeExitError); ok {
-					errCh <- exiterr
-				} else {
-					errCh <- fmt.Errorf("gather never finished for pod %s: %s", pod.Name, err)
-				}
-				return
-			}
+	for _, pod := range pods {
+		podChan <- pod
+	}
+	close(podChan)
 
-			// copy the gathered files into the local destination dir
-			log("downloading gather output")
-			pod, err = o.Client.CoreV1().Pods(pod.Namespace).Get(context.TODO(), pod.Name, metav1.GetOptions{})
-			if err != nil {
-				log("gather output not downloaded: %v\n", err)
-				errCh <- fmt.Errorf("unable to download output from pod %s: %s", pod.Name, err)
-				return
-			}
-			if err := o.copyFilesFromPod(pod); err != nil {
-				log("gather output not downloaded: %v\n", err)
-				errCh <- fmt.Errorf("unable to download output from pod %s: %s", pod.Name, err)
-				return
-			}
-		}(pod)
+	wg.Add(concurrentMG)
+	for i := 0; i < concurrentMG; i++ {
+		go o.runWorker(ns.Name, podChan, errCh, &wg)
```

Comment:
Nit: given the current implementation does not use a queue this could be inlined to avoid passing the arguments:
```
for i := 0; i < concurrentMG; i++ {
	go func() {
		for pod := range podChan {
			if err := o.processNextWorkItem(ns.Name, pod); err != nil {
				errCh <- err
			}
		}
		wg.Done()
	}()
}
```

Reply:
Makes sense. May I ask what's the difference between a queue pattern and worker pool pattern? My web search-fu made me believe that they are different terms for the same thing. Feel free to share a link to RTFM. :)

Reply:
My original intention was to use https://github.com/openshift/library-go/blob/964bcb3f545c24f15294fd8ba914529bf4fe8c4d/pkg/controller/factory/base_controller.go#L246 for popping the pods/items. So in case there's e.g. an error and it make sense to re-queue the item back, we can use https://github.com/openshift/library-go/blob/964bcb3f545c24f15294fd8ba914529bf4fe8c4d/pkg/controller/factory/base_controller.go#L271. Though, this can be added later once we have a case for it.

Reply:
Marking resolved since I have implemented using the suggested approach.